File: packages/capacitor-plugin/android/src/main/java/com/otakit/updater/ManifestKeyConfig.java
package com.otakit.updater;

import android.util.Base64;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONObject;

/** Empty means intentionally unconfigured; malformed explicit settings must never return empty. */
final class ManifestKeyConfig {

  private ManifestKeyConfig() {}

  static List<ManifestVerifier.KeyEntry> parse(JSONObject config) {
    List<ManifestVerifier.KeyEntry> keys = new ArrayList<>();
    try {
      Object raw = config.opt("manifestKeys");
      if (raw == null) return keys;
      if (!(raw instanceof JSONArray)) return invalid();
      JSONArray entries = (JSONArray) raw;
      if (entries.length() == 0) return keys;
      for (int i = 0; i < entries.length(); i++) {
        JSONObject entry = entries.optJSONObject(i);
        if (entry == null) continue;
        String kid = entry.optString("kid", null);
        String key = entry.optString("key", null);
        if (kid != null && key != null) {
          keys.add(new ManifestVerifier.KeyEntry(kid, Base64.decode(key, Base64.DEFAULT)));
        }
      }
      if (keys.isEmpty()) return invalid();
      return keys;
    } catch (Exception error) {
      return invalid();
    }
  }

  private static List<ManifestVerifier.KeyEntry> invalid() {
    android.util.Log.e(
      "OtaKit",
      "Invalid manifestKeys configuration. Manifest verification will reject all updates."
    );
    return java.util.Collections.singletonList(
      new ManifestVerifier.KeyEntry("_invalid_", new byte[0])
    );
  }
}

File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/ManifestKeyConfig.swift
import Foundation

/// Empty means intentionally unconfigured; malformed explicit settings must never return empty.
enum ManifestKeyConfig {
  static func parse(_ config: [String: Any]) -> [(kid: String, key: Data)] {
    guard let rawValue = config["manifestKeys"] else { return [] }
    guard let entries = rawValue as? [[String: String]] else { return invalid() }
    let keys: [(kid: String, key: Data)] = entries.compactMap { entry in
      guard let kid = entry["kid"], let encoded = entry["key"],
            let data = Data(base64Encoded: encoded) else { return nil }
      return (kid: kid, key: data)
    }
    return keys.isEmpty && !entries.isEmpty ? invalid() : keys
  }

  private static func invalid() -> [(kid: String, key: Data)] {
    print("[OtaKit] ERROR: Invalid manifestKeys configuration. Manifest verification will reject all updates.")
    return [(kid: "_invalid_", key: Data())]
  }
}
